### ISSUE TITLE
Deserialize return value

### ISSUE THREAD

**Issue by Rochet2** [bakpakin/binser#2]
This is just some feedback, but from user stand point the vararg return is very inconvenient.

TL;DR
To process the return values from deserialize a new function needs to be created OR must assume there are no nils and they must be tabled like `{binser.deserialize(str)}` to `for` loop them. Also this is complicated more by the fact that deserialize can error out and pcall is needed.

To process the results you must loop them through unless you expect an exact amount of results like 3 and store them in variables.
Looping the results must be done either by `select` or making a table of the results and using `select("#", ...)` to get the amount of results. Doing so requires another function to be made so you can actually use the vararg variable at all `...`.

In addition to this, the serialize and deserialize functions can error out.
Lets assume you want to load game data and want to handle any errors from deserializing.
To do that you would need to wrap the deserialize in a `pcall`, which returns true or false and then error message or the function return value.

Below I made the easiest way to do this, which requires the result of pcall to be directly passed to another function that processes the results. Otherwise I could table all the return values and test if index 1 has true or false and then continue or possibly remove that value.
``` lua
local process_varargs(process, success, ...)
    assert(process, success) -- assert or some other error handler
    for i = 1, select("#", ...) do
        process(select(i, ...))
    end
    return true
end

process_varargs(process, pcall(binser.deserialize, str))
```

If the table and len was returned directly from deserialize without unpacking, it would be very easy to unpack the table if needed and otherwise it would be easy to handle the table and it would also be more efficient. For example
```lua
local success, results, len = pcall(binser.deserialize, str)
assert(success, results) -- assert or some other error handler
for i = 1, len do
    process(results[i])
end
```

**Comment by bakpakin** [comment]
Interesting point. Your suggestion is probably more robust in every possible way. I took the varargs approach for ease of use, but I guess I missed the obvious and important use case. 

```lua
-- Normal, old style
local r1, r2, r2 = binser.deserailize(str)

-- Safe, new style
local success, results, len = pcall(binser.deserialize, str)
assert(success, results)
for i = 1, len do
     print(results[i])
end
```

I will probably update the API to change this.

**Comment by bakpakin** [comment]
I just pushed a change that should hopefully make use easier. `binser.deserialize` now behaves exactly as your desired behavior. Surprisingly, I just had to change one line of code for that.

I also added a new form, `deserializeN(str, n)`, which behaves much like the old form in that it unwraps the values for you. However, you have to pass your explicit maximum number to wrap in n. This hopefully resolves some ambiguity in the simple case, with the added benefit that we can now ignore values we don't need and not waste time parsing them. Nils can still be potentially ambiguous, though.

Unfortunately, this completely breaks the API. Let me know is this works for you, and you can close the issue.

**Comment by Rochet2** [comment]
I think one nice idea could be something like `local tbl, str = binser.deserialize(str, n)` which can be called multiple times to deserialize `n` values from the string. Like the readme said, the data should be concatenable so deserializing should also be possible in parts similarly.


Yeah. Seems like this does the job.
And API breaking is very unfortunate yes. Of course there could be different flavors of deserialization for different uses.
